### PR TITLE
Only include permeability derivatives when required

### PR DESCRIPTION
**Pull request by cpgr**
These calculations can be expensive, so should only
be done when the permeability is not constant.

Refs #13857

### CODE REVIEW

**Comment by cpgr** [comment]
This part of the Jacobian calculation takes up to 20% of the wall time in some of the porous flow heavy tests, even when the derivatives of permeability were zero. This basically adds a flag so that this part of the Jacobian is only calculated when necessary.

For example, on my MacBook `theis_brineco2` takes about 100 s before this change, and 82 s after adding this.

**Review by permcody — APPROVED** [review]

File: modules/porous_flow/include/userobjects/PorousFlowDictator.h [hunk 1]
```diff
@@ -148,6 +148,18 @@ class PorousFlowDictator : public GeneralUserObject, public Coupleable
...
+   * Check if the simulation includes derivatives of permeability
+   * Note: when the permeability is constant, expensive tensor calculations
+   * can be ignored in Jacobian calculations
+   */
+  bool usePermDerivs() const { return _perm_derivs; };
+
+  /**
+   * Set the _perm_derivs flag
+   */
+  void usePermDerivs(bool flag) const { _perm_derivs = flag; };
```
Comment (WilkAndy) [inline modules/porous_flow/include/userobjects/PorousFlowDictator.h]: This whole concept is super nice.  It doesn't make the code crazily complicated, but offers nice speedups.

I really don't the way you've got `usePermDerivs` with multiple signatures!  Quite confusing to my mind, but obviously no one else cared :-)

**Comment by WilkAndy** [comment]
By the way, should we generalise this to other derivs that we don't need all the time?  I'm not sure which ones are expensive, but i'm just asking so we have the idea in our minds that we may have lots of things like `_perm_derivs` sometime in the future.

File: modules/porous_flow/include/userobjects/PorousFlowDictator.h (hunk 1 above)
Comment (cpgr) [reply modules/porous_flow/include/userobjects/PorousFlowDictator.h]: Haha - I started like this, changed it to different function names but as they were quite similar I couldn't remember which was which, so I changed it back

**Comment by cpgr** [comment]
I haven't looked at all the kernels. This one jumped out at me though as it was taking so much time while I was looking at ways to make the brine-co2 class run faster. 

As everything accesses the dictator, it should be easy to do similar things where we can turn off heavy bits of code when they are not needed.

**Pull request merged**